File: sections/62_codeforces_course_ladder/problems/37_xenons_attack/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, q = data[0], data[1]
    size = 1
    while size < n:
        size *= 2

    tree = [-1] * (2 * size)
    tree[size : size + n] = data[2 : 2 + n]
    for node in range(size - 1, 0, -1):
        tree[node] = max(tree[2 * node], tree[2 * node + 1])

    pointer = 2 + n
    output = []
    for _ in range(q):
        operation = data[pointer]
        pointer += 1
        if operation == 1:
            index, capacity = data[pointer], data[pointer + 1]
            pointer += 2
            node = size + index - 1
            tree[node] = capacity
            node //= 2
            while node:
                tree[node] = max(tree[2 * node], tree[2 * node + 1])
                node //= 2
        else:
            need = data[pointer]
            pointer += 1
            if tree[1] < need:
                output.append("0")
                continue
            node = 1
            while node < size:
                if tree[2 * node] >= need:
                    node *= 2
                else:
                    node = 2 * node + 1
            output.append(str(node - size + 1))
    print("\n".join(output))


if __name__ == "__main__":
    main()

File: sections/62_codeforces_course_ladder/problems/37_xenons_attack/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    if (!(cin >> n >> q)) return 0;
    int size = 1;
    while (size < n) size <<= 1;

    vector<long long> tree(2 * size, -1);
    for (int i = 0; i < n; i++) cin >> tree[size + i];
    for (int v = size - 1; v >= 1; v--) {
        tree[v] = max(tree[2 * v], tree[2 * v + 1]);
    }

    while (q--) {
        int type;
        cin >> type;
        if (type == 1) {
            int index;
            long long capacity;
            cin >> index >> capacity;
            int v = size + index - 1;
            tree[v] = capacity;
            for (v >>= 1; v > 0; v >>= 1) {
                tree[v] = max(tree[2 * v], tree[2 * v + 1]);
            }
        } else {
            long long need;
            cin >> need;
            if (tree[1] < need) {
                cout << 0 << '\n';
                continue;
            }
            int v = 1;
            while (v < size) {
                if (tree[2 * v] >= need) v = 2 * v;
                else v = 2 * v + 1;
            }
            cout << v - size + 1 << '\n';
        }
    }
    return 0;
}
